swarm: Add `FromFn` `ConnectionHandler`

thomaseizinger opened the pull request:
## Description

<!-- Please write a summary of your changes and why you made them.-->

The `from_fn` handler is a building block for describing protocols that have little or no connection-specific state. It encourages you to implement your protocol as a function of data without suspending the IO stream. This allows you to use regular `async` functions to express whatever message handshake you want.

To allow sharing of data between connections, `from_fn` uses an abstraction called `Shared`. It implements the observer pattern and pushes new versions of this data to all connections every time it gets updated. Each new stream (inbound or outbound) gets a copy of this state when created.

The stream handlers can either complete with a single event or emit several events as they progress. Thus, this abstraction should also help with the much requested "streaming response" protocol.

With `from_fn`, you still need to implement your own `NetworkBehaviour`. This is considered to be relatively easy though. Not having to implement a `ConnectionHandler` is a major step towards better usability.

## Links to any relevant issues

<!-- Reference any related issues.-->


## Open Questions

<!-- Unresolved questions, if any. -->

- I think we need a better name. Any suggestions?
	- `async_fn_handler`?
- Should AutoNat also be converted?
- Should identify also be converted?
- Should Kademlia be converted? As long as can fit the DHT into memory, this would be possible.
- What do we do with `libp2p-request-response`?

## Open tasks

- [x] Propose `ReadyUpgrade` as separate PR: https://github.com/libp2p/rust-libp2p/pull/2855
- [x] Add a utility for notifying all handlers
- [x] Implement limit for max incoming substreams (const generic vs regular number?)
- [x] Implement limit for max pending dials?
- [ ] Fix TODOs in code
- [x] Experiment with this handler in other protocols (ping? rendezvous? request-response?)

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Add `ReadyUpgrade`

thomaseizinger committed:
Add basic `FromFn` `ConnectionHandler`

thomaseizinger committed:
Add `TState` abstraction

thomaseizinger committed:
Publicly expose types

thomaseizinger committed:
Fix docs

thomaseizinger committed:
Implement limit for max inbound streams and pending dials

thomaseizinger committed:
Implement std::error::Error for `OpenError`

mxinden commented on swarm/src/handler/from_fn.rs:
```diff
@@ -0,0 +1,349 @@
+            + Send,
+    >,
+
+    idle_waker: Option<Waker>,
```
Currently implementations of `ConnectionHandler` depend on being `poll`ed after any of the `inject_xxx` methods are called. I don't think that implicit assumption is great. I prefer the approach you take here, making it explicit via `idle_waker`. The only downside is the additional complexity.

thomaseizinger replied:
Right, I didn't even think about it but of course, with our `loop`s, we implicitly poll components again after we have called functions on them.

`ConnectionHandler` is not going to be used outside of `NetworkBehaviour` so I think it is fine to depend on this and remove the complexity from here.

mxinden commented on swarm/src/handler/from_fn.rs:
```diff
@@ -0,0 +1,349 @@
+            on_new_outbound(stream, state, info).boxed()
+        }),
+        idle_waker: None,
+        pending_dials: VecDeque::default(),
```
```suggestion
        pending_outbound: VecDeque::default(),
```

It took me a bit to understand that this is an outbound stream and not connection. I associate _dial_ with a connection.

thomaseizinger replied:
Makes sense, I'll change it!

mxinden reviewed (commented):
I have to give this abstraction more thought. I find this abstraction to be very similar to `libp2p-request-response`. I have to think about whether it makes sense to have this next to `libp2p-request-response`, have this replace `libp2p-request-response` or have `libp2p-request-response` build on top of this.

elenaf9 commented on swarm/src/handler/from_fn.rs:
```diff
@@ -0,0 +1,349 @@
+where
+    TInboundFuture: Future<Output = TInbound> + Send + 'static,
+    TOutboundFuture: Future<Output = TOutbound> + Send + 'static,
+    TState: Default,
```
What do you think about having a `TState` passed as parameter instead of using `Default`? Use case would be config parameters like timeouts, or for example the next_request_id `Arc` in the request-response protocol:
https://github.com/libp2p/rust-libp2p/blob/89f898c69f6e58944e746565c3f8f8a3bbf3324e/protocols/request-response/src/handler.rs#L95-L96

What I think we could then also do is make `on_new_inbound` and `on_new_outbound` function pointers instead of closures. Anything that they need to capture from the environment can be passed in the initial state.

thomaseizinger replied:
Fair enough. I wanted to reduce the number of parameters but having this configurable seems reasonable!

thomaseizinger replied:
Wouldn't function pointers be unnecessarily limiting? What if the user wants to close over certain variables?

Is your argument that they should capture it in `TState` instead?

elenaf9 replied:
> Is your argument that they should capture it in TState instead?

Yes exactly (provided that we decide to make `TState` a parameter). I don't think we would limit the user with this in any case, and we could reduce complexity a bit. Can you think of any case that would not be possible with this?

thomaseizinger replied:
I think everything would still be possible but it is just not as ergonomic. I'd assume that users are used to closures and having to think how to "emulate" them creates complexity.

Why would it be simpler to have a function pointer?

thomaseizinger replied:
To be clear, I agree making `TState` a parameter but I'd keep the closures.

elenaf9 replied:
> I think everything would still be possible but it is just not as ergonomic. I'd assume that users are used to closures and having to think how to "emulate" them creates complexity.
> 
> Why would it be simpler to have a function pointer?


We would not need to box them and it would improve readability a bit. But I also see your point, so I am fine with sticking with the generic `Fn`.

thomaseizinger replied:
Thanks for the suggestion! The boxing is internal so I don't think it concerns users much. I think the best of both worlds is allowing users to pass in `TState` but also allow closures.

Now that I think of it, we can actually make them `FnMut`s to allow even more use cases.

elenaf9 commented on swarm/src/handler/from_fn.rs:
```diff
@@ -0,0 +1,349 @@
+}
+
+#[derive(Debug)]
+pub enum OutEvent<I, O, OpenInfo> {
```
Related to #2854: rename `OutEvent` -> `ToBehaviour` and `InEvent` -> `FromBehaviour` / `ToHandler` (the latter would describe it better from the behaviour's / Users PoV) ?

thomaseizinger replied:
Unless we agree on https://github.com/libp2p/rust-libp2p/issues/2854, I'd refrain from renaming it here and rather rename them all in one go.

elenaf9 commented on swarm/src/handler/from_fn.rs:
```diff
@@ -0,0 +1,349 @@
+                    outbound_open_info,
+                ),
+            });
+        }
```
Following our coding guidelines i think we should move this block to the bottom so that we first return finished local work (finished upgrades on existing substreams) before we open new substreams.

elenaf9 commented on swarm/src/handler/from_fn.rs:
```diff
@@ -0,0 +1,349 @@
+    on_new_outbound: Box<
+        dyn Fn(NegotiatedSubstream, &mut TState, TOutboundInfo) -> BoxFuture<'static, TOutbound>
+            + Send,
+    >,
```
If we use a separate `FromFnProto` and implement `IntoConnectionHandler`, we would know the `PeerId` and `ConnectedPoint` of the remote and make it an parameter of the user's function. 
So it would then be `Fn(NegociatedSubstream, &mut TState, PeerId, ConnectedPoint)`.
Use case would be e.g. if we have different upgrades depending on whether a connection is relayed not, or depending on the PeerId, etc..

thomaseizinger replied:
That sounds like a good addition!

thomaseizinger replied:
I think this will allow for a really neat implementation of the identify protocol because we just have the observed address available in here :)

elenaf9 reviewed (commented):
👍 I like the general idea. 

> I tried using this locally for rendezvous and we would have to use this mechanism to push the registration data from the behaviour down to the connections.

Did you manage to implement the rendezvous or any other protocol completely with this handler?

> Is the name any good?

No strong opinion. Maybe `FromUpgrade` would describe it a bit better?

> @mxinden @elenaf9 What do you think of adding `NotifyHandler::All`?

I don't see a reason not to add it, and updating the state of all handlers sounds like a valid use-case, so I'm in favor of it.

thomaseizinger committed:
Avoid dial terminology for streams

thomaseizinger committed:
Remove `idle_waker`

thomaseizinger committed:
Make `TState` configurable

thomaseizinger committed:
Finish local work before producing new work

thomaseizinger committed:
Introduce `FromFnProto`

This allows us to capture the `remote_peer_id` and
`connection_point` of the connection.

thomaseizinger committed:
Expose `remote_peer_id` and `connected_point` to closures

thomaseizinger committed:
Merge branch 'master' into from-fn-connection-handler

thomaseizinger committed:
Implement `Shared` abstraction to automatically share state

thomaseizinger committed:
Don't allow ConnectionHandlers to modify the state

It is impossible to implement a general algorithm for how to
resolve conflicts without going down the path of CRDTs.

thomaseizinger committed:
Implement test

thomaseizinger committed:
Remove unnecessary waker

thomaseizinger committed:
Separate clone-able registration data

thomaseizinger committed:
WIP: Migrate `libp2p-rendezvous` to `from_fn`

thomaseizinger commented on protocols/rendezvous/src/server.rs:
```diff
@@ -109,190 +113,169 @@ pub enum Event {
+
+                    None
+                }
+            },
```
This is where we harvest the fruits of this work! Because the `ConnectionHandler` has access to `RegistrationData`, we can "answer" the remote's query inline which allows us to use `async/await` to write the protocol. In the end, it is just a simple request-response but we couldn't do this before because we needed to suspend the stream and wait for the data coming back from the behaviour!

I think this could be a good fit for any protocol where all the data nicely fits into memory.

thomaseizinger replied:
> I think this could be a good fit for any protocol where all the data nicely fits into memory.

Note that with the latest patches, the data is shared between all connections with an `Arc` and thus not actually duplicated in-memory.

thomaseizinger committed:
fixup! WIP: Migrate `libp2p-rendezvous` to `from_fn`

thomaseizinger committed:
fixup! WIP: Migrate `libp2p-rendezvous` to `from_fn`

thomaseizinger committed:
Enforce the use of `Shared`

thomaseizinger committed:
Share state between connection handlers

This reduces the memory usage.

thomaseizinger committed:
Introduce builder pattern

thomaseizinger committed:
Update docs

thomaseizinger committed:
Make inbound streams impossible when no handler is configured

thomaseizinger committed:
Allow for streaming protocols

thomaseizinger committed:
Pass `ConnectedPoint` by value

This clone is not very expensive and offers a huge ergonomic
improvement.

thomaseizinger committed:
Allow regular async functions to be used as handlers

thomaseizinger committed:
Flatten errors

thomaseizinger committed:
Implement `libp2p-ping` with `from_fn` abstraction

thomaseizinger commented on swarm/src/handler/from_fn.rs:
```diff
@@ -0,0 +1,928 @@
+
+    pub fn unregister_connection(&mut self, peer_id: PeerId, id: ConnectionId) {
+        self.connections.remove(&(peer_id, id));
+    }
```
Once https://github.com/libp2p/rust-libp2p/issues/2832 is fixed, we can make this more ergonomic by just requiring a reference to the `FromSwarm` event and we can pick out the data as we need.

thomaseizinger committed:
Merge branch 'master' into from-fn-connection-handler

thomaseizinger committed:
Fully migrate rendezvous

thomaseizinger committed:
Minimise diff

thomaseizinger commented on protocols/rendezvous/src/server.rs:
```diff
@@ -109,192 +149,249 @@ pub enum Event {
+
+                vec![NetworkBehaviourAction::GenerateEvent(
+                    Event::PeerUnregistered { peer, namespace },
+                )]
```
A lot of code here is only present to make things observable from the outside. We don't really need to emit all these events in order for the protocol to be functional.

rkuhn commented on protocols/ping/src/lib.rs:
```diff
@@ -118,30 +232,159 @@ impl NetworkBehaviour for Behaviour {
+                        }
+                    })
+                }
+            })
```
yes, this looks pretty neat!

rkuhn commented on swarm/src/handler/from_fn.rs:
```diff
@@ -0,0 +1,927 @@
+    pub fn with_inbound_handler<TInbound, TInboundStream>(
+        self,
+        max_inbound: usize,
+        handler: impl Fn(NegotiatedSubstream, PeerId, ConnectedPoint, Arc<TState>) -> TInboundStream
```
Please include the protocol name (also for outbound).

thomaseizinger replied:
At the moment, there is only one protocol possible so it will always be that one if that function is called.

rkuhn replied:
I know, and I’d like to [change that](https://github.com/rkuhn/blog/blob/master/08_peer-to-peer_internet.md). All prefabricated upgrades today ignore the negotiated protocol name, which to my mind leaves the second most important value provided by libp2p on the table (the first one being the ability to dial a PeerId).

rkuhn commented on swarm/src/handler/from_fn.rs:
```diff
@@ -0,0 +1,927 @@
+    max_inbound: usize,
+    inbound_handler: Option<
+        Box<
+            dyn Fn(
```
AFAICS this restricts the part of the handler that constructs the stream processor to shared access to its closure — is there a reason for this? If not I’d prefer FnMut.

thomaseizinger replied:
No particular reason, I guess this could be useful if you want to track connection-local state across all inbound/outbound streams?

That would only be shared across one closure though, meaning you wouldn't be able to share state between the inbound and outbound factory function.

I think we might be able to change this interface to allow for a single factory function that gets called for both inbound and outbound streams which would allow you to share state. Is that what you would be after?

rkuhn replied:
Sharing state is not my concern because doing that always proceeds via shared references (of which Arc is a special case). `Fn` does not allow captured values to be modified (more precisely: mutably borrowed), even if they are owned.

thomaseizinger replied:
> `Fn` does not allow captured values to be modified (more precisely: mutably borrowed), even if they are owned.

If you want to mutate a captured value, you are effectively sharing the state within a connection across all stream handler factory invocations.

I think we are saying the same thing and I agree that this capability is useful :)

rkuhn replied:
Ah, you mean “sharing along the timeline”, not “sharing in different places”. Yes, now we both know that we are in agreement :-)

rkuhn commented on swarm/src/handler/from_fn.rs:
```diff
@@ -0,0 +1,927 @@
+/// state changes.
+///
+/// [`NetworkBehaviour`]: crate::NetworkBehaviour
+pub fn from_fn(protocol: &'static str) -> Builder<WantState> {
```
please allow multiple protocols, like
```rust
pub fn from_fn<I>(protocols: I) -> Builder<I, WantState>
where
    I: IntoIterator + Clone + Send + 'static
    I::Item: ProtocolName
{ ... }
```

thomaseizinger replied:
Yeah that is doable!

rkuhn commented on swarm/src/handler/from_fn.rs:
```diff
@@ -0,0 +1,927 @@
+        {
+            if self.keep_alive.is_yes() {
+                // TODO: Make timeout configurable
+                self.keep_alive = KeepAlive::Until(Instant::now() + Duration::from_secs(10))
```
It feels weird that this grace period is implemented on the level of ConnectionHandler — shouldn’t it be a NetworkBehaviour setting that applies whenever KeepAlive toggles from Yes to No?

thomaseizinger replied:
I am not happy with this either but I couldn't think of a better way to handle this. `keep_alive` is per connection. My initial implementation would have been a `with_keep_alive_timeout` configuration parameter on the builder that defaults to 30s or something.

rkuhn commented on swarm/src/handler/from_fn.rs:
```diff
@@ -0,0 +1,927 @@
+                    PeerId,
+                    ConnectedPoint,
+                    Arc<TState>,
+                ) -> BoxStream<'static, TInbound>
```
I’d prefer type aliases for these function types

thomaseizinger replied:
What do you mean exactly? As a user, you shouldn't need to come in contact this this `BoxStream` per se. Where would you like to see a type alias?

rkuhn replied:
I only mean the readability of this source module by itself.

rkuhn commented on swarm/src/handler/from_fn.rs:
```diff
@@ -0,0 +1,927 @@
+                .connections
+                .iter()
+                .map(|(peer_id, conn_id)| (*peer_id, *conn_id, self.shared.clone()))
+                .collect();
```
This erases previously undelivered wake-ups — which is fine unless it happens in a tight loop.

thomaseizinger replied:
The old events are useless though because they contain stale state. I am not sure if there is a better way?

rkuhn replied:
Right, old events are useless. Perhaps I am overthinking this, my worry is that there may be a situation where only the first (few) connections are ever notified because updates keep coming in — a fairness concern. This would be solved by randomising the notification order, or by always finishing one batch of notifications (with values updated during the process!) before starting the next. The second is deterministic but more work to implement.

Getting these things wrong usually has low chance of severe impact, but when the impact happens it is really difficult for downstream code to find the bug.

thomaseizinger replied:
I think that we currently drain events from the `NetworkBehaviour` before we inject new ones into it so I don't think that this would happen but I'd have to double check that.

rkuhn reviewed (commented):
Yes, this does indeed look quite promising!

thomaseizinger committed:
Merge branch 'master' into from-fn-connection-handler

thomaseizinger committed:
Update to new behaviour interface
